### PR TITLE
Find best route

### PR DESCRIPTION
**Pull request by siarheishut**
Add config ```RouterSettings``` for ```BusManager``` and functionality for finding the best in terms of time route between any two stops.

### CODE REVIEW

File: src/request_types.h [hunk 1]
```diff
@@ -9,6 +9,11 @@
 #include <map>
 
 namespace rm {
+struct RoutingSettings {
+  int bus_wait_time;
+  double bus_velocity;
```
Comment (foxfrikses) [inline src/request_types.h:14]: Could you add a comment specifying in which units `bus_wait_time` and `bus_velocity` are expected to be specified?

File: src/route_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
+#include "route_manager.h"
+#include "distance_computer.h"
+#include "common.h"
+
+#include <variant>
```
Comment (foxfrikses) [inline src/route_manager.cpp:5]: Missing includes: `memory`, `vector`, `optional`, `string`, `unordered_map`

File: src/route_manager.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#include "request_types.h"
+#include "router.h"
+#include "common.h"
+
+#include <optional>
+#include <memory>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+#include <utility>
```
Comment (foxfrikses) [inline src/route_manager.h:13]: What `utility` is for?

File: src/main.cpp [hunk 4]
```diff
@@ -13,20 +13,24 @@ int main() {
 
   auto base_requests = input_map.find("base_requests");
   auto stat_requests = input_map.find("stat_requests");
+  auto routing_settings = input_map.find("routing_settings");
 
   if (base_requests == input_map.end()
-      || stat_requests == input_map.end())
+      || stat_requests == input_map.end()
+      || routing_settings == input_map.end())
```
Comment (foxfrikses) [inline src/main.cpp:20]: `||` to the end of the line

File: src/request_types.h [hunk 5]
```diff
@@ -26,8 +26,7 @@ struct GetStopRequest {
 
 struct FindRouteRequest {
   int id;
-  std::string from;
-  std::string to;
+  std::string from, to;
```
Comment (foxfrikses) [inline src/request_types.h:29]: I feel like this should be already reflected in the first commit, which added this struct.

But actually I think that we shouldn't declare two fields in the same row. It damages readability.

File: tests/bus_manager_test.cpp [hunk 6]
```diff
@@ -455,3 +477,118 @@ TEST(TestBusManager, TestGetStopInfo) {
...
+                  .from = "stop 1",
+                  .to = "stop 2"},
+              FindRouteRequest{
+                  .id = 2,
+                  .from = "stop 2",
+                  .to = "stop 3"},
+              FindRouteRequest{
+                  .id = 3,
+                  .from = "stop 3",
+                  .to = "stop 4"}
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:512]: One/two would be enough

File: tests/bus_manager_test.cpp [hunk 7]
```diff
@@ -455,3 +477,118 @@ TEST(TestBusManager, TestGetStopInfo) {
...
+          },
+          .want = {std::nullopt,
+                   std::nullopt,
+                   std::nullopt},
+      },
+      TestCase{
+          .name = "Common requests",
```
Comment (foxfrikses) [inline tests/bus_manager_test.cpp:519]: I would add requests for an existing database:
- request an unachievable path (between two different components (graph)).
- request a path between an existing stop and a missing one.

File: tests/request_parser_test.cpp [hunk 8]
```diff
@@ -320,7 +320,58 @@ TEST(TestOutputRequest, TestGetStopRequest) {
...
+          .name = "Wrong request: no <id>",
+          .input = json::Dict{{"type", "Bus"},
+                              {"from", "stop 1"},
+                              {"to", "stop 2"}},
+          .want = std::nullopt,
+      },
+      TestCase{
+          .name = "Common request",
+          .input = json::Dict{{"type", "Route"},
+                              {"from", "stop 1"},
```
Comment (foxfrikses) [inline tests/request_parser_test.cpp:358]: Wrong type: `id` is `double`, `from` is `int`, `to` is `double`

File: tests/request_parser_test.cpp [hunk 9]
```diff
@@ -320,7 +320,58 @@ TEST(TestOutputRequest, TestGetStopRequest) {
...
+      },
+  };
+
+  for (auto &[name, input, want] : test_cases) {
+    auto got = ParseFindRouteRequest(input);
+    EXPECT_EQ(want, got) << name;
+  }
+}
+
+TEST(TestParseRequestP, TestOutput) {
```
Comment (foxfrikses) [inline tests/request_parser_test.cpp:374]: Test `ParseSettings`?

File: src/request_processor.cpp [hunk 10]
```diff
@@ -4,6 +4,40 @@
 #include <variant>
 
 namespace rm {
+json::List ToJson(std::vector<RouteInfo::Item> response_items) {
+  json::List items;
+  for (auto &item : response_items) {
+    if (std::holds_alternative<RouteInfo::WaitItem>(item)) {
+      auto ptr_w = std::get<RouteInfo::WaitItem>(item);
```
Comment (foxfrikses) [inline src/request_processor.cpp:11]: It's not a prt, but a ref.

`auto& w = std::get<RouteInfo::WaitItem>(item);`

Same with `ptr_r`.

File: src/request_processor.cpp [hunk 11]
```diff
@@ -4,6 +4,40 @@
 #include <variant>
 
 namespace rm {
+json::List ToJson(std::vector<RouteInfo::Item> response_items) {
```
Comment (foxfrikses) [inline src/request_processor.cpp:7]: Please use `RouteResponse` instead of `RouteInfo` in this file.

File: src/request_processor.h [hunk 12]
```diff
@@ -8,11 +8,14 @@
 #include <ostream>
 
 namespace rm {
+json::List ToJson(std::vector<RouteInfo::Item> response_items);
```
Comment (foxfrikses) [inline src/request_processor.h:11]: Please use `RouteResponse` instead of `RouteInfo` in this file.

File: tests/request_processor_test.cpp [hunk 13]
```diff
@@ -86,3 +86,53 @@ TEST(TestProcessRequests, TestBusResponseToJson) {
...
+  std::vector<TestCase> test_cases{
+      TestCase{
+          .name = "No route",
+          .response = std::nullopt,
+          .id = 1234567,
+          .want = json::Dict{{"request_id", 1234567},
+                             {"error_message", "not found"}},
+      },
+      TestCase{
+          .name = "Bus Found",
```
Comment (foxfrikses) [inline tests/request_processor_test.cpp:109]: Route found

File: tests/request_processor_test.cpp [hunk 14]
```diff
@@ -86,3 +86,53 @@ TEST(TestProcessRequests, TestBusResponseToJson) {
...
+          .response = rm::RouteResponse{
+              .time = 25.135,
+              .items = {
+                  rm::RouteResponse::WaitItem{
+                      .stop = "Stop 1",
+                      .time = 12},
+                  rm::RouteResponse::RoadItem{
+                      .bus = "Bus 1",
+                      .time = 13.135,
+                      .span_count = 3},
```
Comment (foxfrikses) [inline tests/request_processor_test.cpp:119]: Let's add a few more items

File: src/request_processor.cpp (hunk 11 above)
Comment (siarheishut) [reply src/request_processor.cpp:7]: Done.

File: src/request_processor.cpp (hunk 10 above)
Comment (siarheishut) [reply src/request_processor.cpp:11]: Changed.

File: src/request_processor.h (hunk 12 above)
Comment (siarheishut) [reply src/request_processor.h:11]: Done.

File: tests/request_processor_test.cpp (hunk 13 above)
Comment (siarheishut) [reply tests/request_processor_test.cpp:109]: Changed.

File: tests/request_processor_test.cpp (hunk 14 above)
Comment (siarheishut) [reply tests/request_processor_test.cpp:119]: Done.

File: tests/request_parser_test.cpp (hunk 9 above)
Comment (siarheishut) [reply tests/request_parser_test.cpp]: Done.

File: tests/request_parser_test.cpp (hunk 8 above)
Comment (siarheishut) [reply tests/request_parser_test.cpp:358]: What does it mean?

**Pull request closed without merging**